text/0000-slatepack.md

```diff
@@ -0,0 +1,391 @@
+
+### Armor
+
+The payload that will be armored is a binary serialized `SlatePack` JSON object. Armor is `Framing` wrapped around a `SimpleBase58Check` encoded `Payload`.
+
+#### Framing
+
+Armor uses specific `Headers`, `Footers` and `Periods` as `Framing` to contain its `Payload`.
+
+- `Header`
+  - Supported Headers:
+    - `BEGINSLATEPACK`
+    - `ENDSLATEPACK X/Y`
+      - `Y` is the total number of message parts for a given multipart message and `X` is the number of the current partial message
+      - Only used if there is more than one message part
+  - Regex: `^[>\n\r\t ]*BEGINSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$`
+
+
+- `Footer`
+  - Supported Footers
+    - `ENDSLATEPACK`
+    - `ENDSLATEPACK X/Y`
+      - `Y` is the total number of message parts for a given multipart message and `X` is the number of the current partial message
+      - Only used if there is more than one message part
+  - Regex: `^[>\n\r\t ]*ENDSLATEPACK[>\n\r\t ]*([1-9]\/[1-9])*$`
+
+- `Periods`
+
+  - All data of an armored slate up to the first `.` is the framing header
+  - All data after the first `.` and before the second `.` is the `SimpleBase58Check` encoded payload which contains the slate data
+  - All data after the second `.` and before the third `.` is the framing footer
+  - Any data after the third `.` is ignored
+
+
+#### `SimpleBase58Check` Encoding
+
+`SlatePack Message` armor payloads are encoded similar to legacy bitcoin addresses, with the primary differences being that the `SimpleBase58Check` used here does not include version bytes and includes the error checking code at the beginning of the payload instead of at the end.
+
+1. `SHA256(SHA256(SLATEPACK_MESSAGE_BINARY))`
+
+2. First four bytes from previous step are `ERROR_CHECK_CODE`
+
+3. Concatenate `ERROR_CHECK_CODE + SLATEPACK_MESSAGE_BINARY`
+
+4. Base58 encode the output from the previous step to complete the armor `Payload`
+
+It should be noted that the `ERROR_CHECK_CODE` does not have a robust error checking ability because a double sha256 hash is not a proper error check code and the encoding scheme itself was meant to be used for bitcoin addresses which are much smaller than slate payloads.
+
+A more robust error correction option was not chosen here because the consequences of the failure to detect an error are not as severe as they would be for a bitcoin address as further validation would need to occur for Grin. The purpose is to catch some characters being accidentally added or lost during armor transport rather than preventing a spend to an address we don't know the key to spend from.
+
+#### Formatting
+
+- `WORD_LENGTH`: `15`
+  - Number of `SimpleBase58Check` encoded characters per word; chosen for human readability across device screen sizes
+
+- `LINE_BREAK`: `200` words
+  - Number of words of `WORD_LENGTH` to include before inserting a newline; chosen for user friendliness in terminals and messaging applications
```

Comment:
Is 15*200=3000 characters per line user friendly? That's a single line for the average 1 output slate. And if we're allowing multiple lines, I can't envision a scenario that would have multiple 10+KB messages which wouldn't be better off just using synchronous transactions. Do we really want to support the complexity of multiple messages?

Reply:
Good question. This parameter selection was based on https://saltpack.org/armoring. I agree that once we get into multipart messages the UX/DX immediately falls off. Ideally in these cases synchronous methods would always be used. 

However if we expect slatepack to be a default standard and if Tor is blocked at packet level in some regions, we still need the fallback method to be able to handle almost any possible case- otherwise a user unable to use Tor that wants to spend a large tx might have trouble.